### DISCUSSION TITLE
Is there a way to force tier 1 compilation of a method?

### DISCUSSION THREAD

**Discussion by ladeak** [dotnet/runtime#95847] [Q&A] [answered]
I have a method in .NET8 (windows, not R2R) that is compiled as (using DOTNET_JitDisasm):

```
; Assembly listing for method MyType:SomeMethod[System.__Canon](System.Collections.Generic.Dictionary`2[System.String,System.String],System.String):System.__Canon (Tier0)
; Emitting BLENDED_CODE for X64 with AVX - Windows
; Tier0 code
; rbp based frame
; partially interruptible
````
Regardless, if I invoke it thousands of times, I do not see a Tier 1 compilation created for it (and this is the only method I have in my solution. Is there a way to gain understanding why there is no Tier 1 compilation happening?

Applying the `AggressiveOptimization` attribute still produces Tier-0 optimized flag.

Is there a way to force tier 1 compilation of a method?

I wonder because I wonder what happens when I use this method in a real production app. Is it not ever going to have a tier 1 code?

**Comment by gfoidl** [comment]
> if I invoke it thousands of times, I do not see a Tier 1 compilation created for it

Tiered compilation kicks in when the method is called often enough and after a period of time where no JIT-compilations happens (to settle a starting phase, etc.).

So when you just call the method at app entry the thousands time the second criteria (time) isn't fulfilled.
Thus try it that way:
* call your method say 50 times
* `Thread.Sleep(250)` to let the statup timeout pass and a bit of time to let the JIT produce Tier-1 code
* call your method again -- now Tier-1 code should be logged via disams

---

Does you code contain loops?

**Reply by ladeak** [reply]
That is exactly what I have. I am looking for `ByTypeCastMediaTypeHeaderValue`:

```csharp
var myObj = ...
int length = 0;
for (int i = 0; i < 16_000; i++)
    length += myObj.ByTypeCastMediaTypeHeaderValue().MediaType.Length;

await Task.Delay(1000);

for (int i = 0; i < 300; i++)
    length += myObj.ByTypeCastMediaTypeHeaderValue().MediaType.Length;

Console.WriteLine(length);
```

**Reply by KeterSCP** [reply, marked as answer]
Can you try with?

```csharp
for (int i = 0; i < 100; i++)
{
    length += myObj.ByTypeCastMediaTypeHeaderValue().MediaType.Length;
    Thread.Sleep(10);
}
```

**Reply by ladeak** [reply]
Yes, that does the trick:

 ```
; Assembly listing for method MyType:SomeMethod[System.__Canon](System.Collections.Generic.Dictionary`2[System.String,System.String],System.String):System.__Canon (Tier1)
; Emitting BLENDED_CODE for X64 with AVX - Windows
; Tier1 code
; optimized code
; optimized using Dynamic PGO
```

**Comment by MichalPetryka** [comment]
> Applying the `AggressiveOptimization` attribute still produces Tier-0 optimized flag.

`AggressiveOptimization` won't emit Tier1 codegen (it'll be slightly worse) but it won't ever emit Tier0, you must've done something wrong to see that.

**Reply by ladeak** [reply]
I recall seeing tier 0 optimized, but might remembering it incorrectly.
